mdsuite/graph_modules/molecular_graph.py

```diff
@@ -60,143 +70,141 @@ def __init__(
 
-    def build_smiles_graph(self):
-        """
-        Build molecular graphs from SMILES strings.
         """
-        mol = read_smiles(self.smiles_string, explicit_hydrogen=True)
-        data = mol.nodes
-        species = {}
-        for i in range(len(data)):
-            item = data[i].get("element")
-            if item in species:
-                species[item] += 1
-            else:
-                species[item] = 0
-                species[item] += 1
+        self.experiment = experiment
+        self.molecule_name = molecule_input_data.name
+        self.database = Database(self.experiment.database_path / "database.hdf5")
+        self.cutoff = molecule_input_data.cutoff
+        self.n_molecules = molecule_input_data.amount
+        self.mol_pbc = molecule_input_data.mol_pbc
 
-        return mol, species
+        if self.mol_pbc:
+            self.reference_property = "Positions"
+        else:
+            self.reference_property = "Unwrapped_Positions"
 
-    @staticmethod
-    def _apply_system_cutoff(tensor: tf.Tensor, cutoff: float) -> tf.Tensor:
+        if isinstance(molecule_input_data.reference_configuration, int):
+            self.reference_configuration = molecule_input_data.reference_configuration
+        else:
+            self.reference_configuration = 0
+
+        if isinstance(molecule_input_data.smiles, str):
+            self.smiles_graph, self.species = build_smiles_graph(
+                molecule_input_data.smiles
+            )
+        elif isinstance(molecule_input_data.species_dict, dict):
+            self.species = molecule_input_data.species_dict
+            self.smiles_string = None
+        else:
+            error_msg = (
+                "The minimum amount of data was not given to the mapping."
+                "Either provide a reference key with information about"
+                "Which species and the number of them are in the molecule,"
+                "or provide a SMILES string that can be used to compute "
+                "this information."
+            )
+            raise ValueError(error_msg)
+
+        self._get_molecular_mass()
+        self._build_molecule_groups()  # populate the class group attribute
+        self._perform_isomorphism_tests()  # run the graph tests.
+
+    def _get_molecular_mass(self):
         """
-        Enforce a cutoff on a tensor
+        Get the mass of a SMILES molecule based on experiment data.
 
-        Parameters
-        ----------
-        tensor : tf.Tensor
-        cutoff : float
-        """
+        Returns
+        -------
+        Updates the following class attributes:
 
-        cutoff_mask = tf.cast(
-            tf.less(tensor, cutoff), dtype=tf.int16
-        )  # Construct the mask
+        mass : float
+                mass of the molecule
+        """
+        mass = 0.0
+        for item in self.species:
+            mass += self.experiment.species[item]["mass"][0] * self.species[item]
```

Comment:
I prefer the one liners when ever I see, that a variable is initialized with 0.0
`mass = sum([self.experiment.species[x]["mass"][0] * self.species[x] for x in self.species])`

Also why is ist `["mass"][0]` ? What are the other dimensions?

Reply:
also what is ``self.species[item]`` and why is it multiplied on the mass?

Reply:
This dates back to when all species properties where stored in lists, mass still being one atm. It is due to, for example in the case of charge be able to store both a single value for the particle or one for each atom. In the case of mass, the initial motivation was that a user could store several masses for one particle (e.g isotopes) and perform computations with each as they are mostly used in pre-factors. As this is impossible and users would have to set that mass manually this can be removed. It will need to be weeded out of the file reading though and so I would recommend a different PR.

Reply:
Seeing as long one liners can make code harder to read and are generally discouraged, and the code already here is very clear and the same as the one-liner, I will not change this, but feel free to change it yourself if it makes you more comfortable.

I have reduced is slightly by moving the unnecessary attribute update and making all of it in reference to `self.molecular_mass`

Reply:
> To answer your question @christophlohrmann self.species[item] is the number of particles of species item in the molecule.

In that case I would suggest renaming it to something that involves the word "number". Especially because in the same line, a variable also called species is a completely different thing

Reply:
I think resolved in b8747e9